File: frontend/src/lib/rbac.ts
/**
 * Client-side mirror of backend/middleware/rbac.py.
 *
 * This is a UI affordance only — it hides controls a role cannot use so the
 * operator is not offered actions that will 403. The backend remains the sole
 * authority; every guarded route re-checks server-side. Keep this table in sync
 * when ROLE_PERMISSIONS changes.
 */

export const Permissions = {
  TELEMETRY_READ: "telemetry.read",
  TELEMETRY_INGEST: "telemetry.ingest",
  INCIDENT_READ: "incident.read",
  INCIDENT_ACKNOWLEDGE: "incident.acknowledge",
  INCIDENT_ASSIGN: "incident.assign",
  INCIDENT_RESOLVE: "incident.resolve",
  INCIDENT_CLOSE: "incident.close",
  DRONE_READ: "drone.read",
  DRONE_COMMAND_REQUEST: "drone.command.request",
  DRONE_COMMAND_APPROVE: "drone.command.approve",
  AUDIT_READ: "audit.read",
  ORGANIZATION_MANAGE: "organization.manage",
  USER_MANAGE: "user.manage",
  GEOFENCE_READ: "geofence.read",
  GEOFENCE_MANAGE: "geofence.manage",
  SETTINGS_READ: "settings.read",
  SETTINGS_MANAGE: "settings.manage",
  AI_EXPLAIN: "ai.explain",
} as const;

export type Permission = (typeof Permissions)[keyof typeof Permissions];

const P = Permissions;

export const ROLE_PERMISSIONS: Record<string, ReadonlySet<Permission>> = {
  admin: new Set<Permission>([
    P.TELEMETRY_READ, P.TELEMETRY_INGEST,
    P.INCIDENT_READ, P.INCIDENT_ACKNOWLEDGE, P.INCIDENT_ASSIGN, P.INCIDENT_RESOLVE, P.INCIDENT_CLOSE,
    P.DRONE_READ, P.DRONE_COMMAND_REQUEST, P.DRONE_COMMAND_APPROVE,
    P.AUDIT_READ, P.ORGANIZATION_MANAGE, P.USER_MANAGE,
    P.GEOFENCE_READ, P.GEOFENCE_MANAGE,
    P.SETTINGS_READ, P.SETTINGS_MANAGE,
    P.AI_EXPLAIN,
  ]),
  commander: new Set<Permission>([
    P.TELEMETRY_READ, P.TELEMETRY_INGEST,
    P.INCIDENT_READ, P.INCIDENT_ACKNOWLEDGE, P.INCIDENT_ASSIGN, P.INCIDENT_RESOLVE, P.INCIDENT_CLOSE,
    P.DRONE_READ, P.DRONE_COMMAND_REQUEST, P.DRONE_COMMAND_APPROVE,
    P.AUDIT_READ,
    P.GEOFENCE_READ, P.GEOFENCE_MANAGE,
    P.SETTINGS_READ,
    P.AI_EXPLAIN,
  ]),
  analyst: new Set<Permission>([
    P.TELEMETRY_READ,
    P.INCIDENT_READ, P.INCIDENT_ACKNOWLEDGE, P.INCIDENT_ASSIGN, P.INCIDENT_RESOLVE,
    P.DRONE_READ,
    P.AUDIT_READ,
    P.GEOFENCE_READ,
    P.SETTINGS_READ,
    P.AI_EXPLAIN,
  ]),
  operator: new Set<Permission>([
    P.TELEMETRY_READ, P.TELEMETRY_INGEST,
    P.INCIDENT_READ,
    P.DRONE_READ, P.DRONE_COMMAND_REQUEST,
    P.GEOFENCE_READ,
    P.SETTINGS_READ,
    P.AI_EXPLAIN,
  ]),
  observer: new Set<Permission>([
    P.TELEMETRY_READ,
    P.INCIDENT_READ,
    P.DRONE_READ,
    P.GEOFENCE_READ,
    P.SETTINGS_READ,
  ]),
};

/** Roles are compared case-insensitively, matching the backend. */
export function hasPermission(role: string | null | undefined, permission: Permission): boolean {
  if (!role) return false;
  return ROLE_PERMISSIONS[role.toLowerCase()]?.has(permission) ?? false;
}

export function permissionsFor(role: string | null | undefined): Permission[] {
  if (!role) return [];
  return [...(ROLE_PERMISSIONS[role.toLowerCase()] ?? [])];
}

/** Roles the router allows into /admin and /settings. Mirrors router.tsx. */
export const ELEVATED_ROLES = ["admin", "commander"] as const;

export function isElevated(role: string | null | undefined): boolean {
  return !!role && (ELEVATED_ROLES as readonly string[]).includes(role.toLowerCase());
}

File: backend/middleware/rbac.py
"""
RBAC Permission Map for SwarmGuard AI Enterprise.

Defines granular permissions and maps them to roles.
"""

# --- Permission Definitions ---
class Permissions:
    # Telemetry
    TELEMETRY_READ = "telemetry.read"
    TELEMETRY_INGEST = "telemetry.ingest"

    # Incidents
    INCIDENT_READ = "incident.read"
    INCIDENT_ACKNOWLEDGE = "incident.acknowledge"
    INCIDENT_ASSIGN = "incident.assign"
    INCIDENT_RESOLVE = "incident.resolve"
    INCIDENT_CLOSE = "incident.close"

    # Drones
    DRONE_READ = "drone.read"
    DRONE_COMMAND_REQUEST = "drone.command.request"
    DRONE_COMMAND_APPROVE = "drone.command.approve"
    # Issue and revoke per-drone ingest keys. Admin only: a key authenticates
    # telemetry as that drone, so issuing one is granting the ability to
    # speak for an aircraft.
    DEVICE_CREDENTIALS_MANAGE = "device.credentials.manage"

    # Audit
    AUDIT_READ = "audit.read"

    # Organization & Users
    ORGANIZATION_MANAGE = "organization.manage"
    USER_MANAGE = "user.manage"

    # Geofence
    GEOFENCE_READ = "geofence.read"
    GEOFENCE_MANAGE = "geofence.manage"

    # Settings
    SETTINGS_READ = "settings.read"
    SETTINGS_MANAGE = "settings.manage"

    # AI
    AI_EXPLAIN = "ai.explain"


# --- Role -> Permission Mapping ---
ROLE_PERMISSIONS: dict[str, set[str]] = {
    "admin": {
        Permissions.TELEMETRY_READ, Permissions.TELEMETRY_INGEST,
        Permissions.INCIDENT_READ, Permissions.INCIDENT_ACKNOWLEDGE,
        Permissions.INCIDENT_ASSIGN, Permissions.INCIDENT_RESOLVE,
        Permissions.INCIDENT_CLOSE,
        Permissions.DRONE_READ, Permissions.DRONE_COMMAND_REQUEST,
        Permissions.DRONE_COMMAND_APPROVE,
        Permissions.AUDIT_READ,
        Permissions.ORGANIZATION_MANAGE, Permissions.USER_MANAGE,
        Permissions.DEVICE_CREDENTIALS_MANAGE,
        Permissions.GEOFENCE_READ, Permissions.GEOFENCE_MANAGE,
        Permissions.SETTINGS_READ, Permissions.SETTINGS_MANAGE,
        Permissions.AI_EXPLAIN,
    },
    "commander": {
        Permissions.TELEMETRY_READ, Permissions.TELEMETRY_INGEST,
        Permissions.INCIDENT_READ, Permissions.INCIDENT_ACKNOWLEDGE,
        Permissions.INCIDENT_ASSIGN, Permissions.INCIDENT_RESOLVE,
        Permissions.INCIDENT_CLOSE,
        Permissions.DRONE_READ, Permissions.DRONE_COMMAND_REQUEST,
        Permissions.DRONE_COMMAND_APPROVE,
        Permissions.AUDIT_READ,
        Permissions.GEOFENCE_READ, Permissions.GEOFENCE_MANAGE,
        Permissions.SETTINGS_READ,
        Permissions.AI_EXPLAIN,
    },
    "analyst": {
        Permissions.TELEMETRY_READ,
        Permissions.INCIDENT_READ, Permissions.INCIDENT_ACKNOWLEDGE,
        Permissions.INCIDENT_ASSIGN, Permissions.INCIDENT_RESOLVE,
        Permissions.DRONE_READ,
        Permissions.AUDIT_READ,
        Permissions.GEOFENCE_READ,
        Permissions.SETTINGS_READ,
        Permissions.AI_EXPLAIN,
    },
    "operator": {
        Permissions.TELEMETRY_READ, Permissions.TELEMETRY_INGEST,
        Permissions.INCIDENT_READ,
        Permissions.DRONE_READ, Permissions.DRONE_COMMAND_REQUEST,
        Permissions.GEOFENCE_READ,
        Permissions.SETTINGS_READ,
        Permissions.AI_EXPLAIN,
    },
    "observer": {
        Permissions.TELEMETRY_READ,
        Permissions.INCIDENT_READ,
        Permissions.DRONE_READ,
        Permissions.GEOFENCE_READ,
        Permissions.SETTINGS_READ,
    },
}


def get_permissions_for_role(role: str) -> set[str]:
    """Returns the set of permissions for a given role."""
    return ROLE_PERMISSIONS.get(role.lower(), set())


def has_permission(role: str, permission: str) -> bool:
    """Checks if a role has a specific permission."""
    return permission in get_permissions_for_role(role)
